# Introduction

Hopefully, this short tutorial can show you a lot of different commands that will help you gain the most insights into your dataset.

In [ ]:
def load_data_from_google_drive(url):
    url_processed='https://drive.google.com/uc?id=' + url.split('/')[-2]
    df = pd.read_csv(url_processed)
    return df

In [ ]:
import pandas as pd

# Loading in Data

The first step in any ML problem is identifying what format your data is in, and then loading it into whatever framework you're using. For Kaggle compeitions, a lot of data can be found in CSV files, so that's the example we're going to use.

We're going to be looking at a sports dataset that shows the results from NCAA basketball games from 1985 to 2016. This dataset is in a CSV file, and the function we're going to use to read in the file is called **pd.read_csv()**. This function returns a **dataframe** variable. The dataframe is the golden jewel data structure for Pandas. It is defined as "a two-dimensional size-mutable, potentially heterogeneous tabular data structure with labeled axes (rows and columns)".

Just think of it as a table for now.

In [ ]:
df = load_data_from_google_drive(url='https://drive.google.com/file/d/184JcLbSpArA_uq0DgAv2k892KChJVPHt/view?usp=share_link')

In [ ]:
df

,Season,Daynum,Wteam,Wscore,Lteam,Lscore,Wloc,Numot
0,1985,20,1228,81,1328,64,N,0
1,1985,25,1106,77,1354,70,H,0
2,1985,25,1112,63,1223,56,H,0
3,1985,25,1165,70,1432,54,H,0
4,1985,25,1192,86,1447,74,H,0
...,...,...,...,...,...,...,...,...
145284,2016,132,1114,70,1419,50,N,0
145285,2016,132,1163,72,1272,58,N,0
145286,2016,132,1246,82,1401,77,N,1
145287,2016,132,1277,66,1345,62,N,0


# The Basics

Now that we have our dataframe in our variable df, let's look at what it contains. We can use the function **head()** to see the first couple rows of the dataframe (or the function **tail()** to see the last few rows).

In [ ]:
df.head()

,Season,Daynum,Wteam,Wscore,Lteam,Lscore,Wloc,Numot
0,1985,20,1228,81,1328,64,N,0
1,1985,25,1106,77,1354,70,H,0
2,1985,25,1112,63,1223,56,H,0
3,1985,25,1165,70,1432,54,H,0
4,1985,25,1192,86,1447,74,H,0


In [ ]:
df.tail()

,Season,Daynum,Wteam,Wscore,Lteam,Lscore,Wloc,Numot
145284,2016,132,1114,70,1419,50,N,0
145285,2016,132,1163,72,1272,58,N,0
145286,2016,132,1246,82,1401,77,N,1
145287,2016,132,1277,66,1345,62,N,0
145288,2016,132,1386,87,1433,74,N,0


We can see the dimensions of the dataframe using the the **shape** attribute

In [ ]:
df.shape

(145289, 8)

We can also extract all the column names as a list, by using the **columns** attribute and can extract the rows with the **index** attribute

In [ ]:
df.columns.tolist()

['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot']

In order to get a better idea of the type of data that we are dealing with, we can call the **describe()** function to see statistics like mean, min, etc about each column of the dataset.

In [ ]:
df.describe()

,Season,Daynum,Wteam,Wscore,Lteam,Lscore,Numot
count,145289.000000,145289.000000,145289.000000,145289.000000,145289.000000,145289.000000,145289.000000
mean,2001.574834,75.223816,1286.720646,76.600321,1282.864064,64.497009,0.044387
std,9.233342,33.287418,104.570275,12.173033,104.829234,11.380625,0.247819
min,1985.000000,0.000000,1101.000000,34.000000,1101.000000,20.000000,0.000000
25%,1994.000000,47.000000,1198.000000,68.000000,1191.000000,57.000000,0.000000
50%,2002.000000,78.000000,1284.000000,76.000000,1280.000000,64.000000,0.000000
75%,2010.000000,103.000000,1379.000000,84.000000,1375.000000,72.000000,0.000000
max,2016.000000,132.000000,1464.000000,186.000000,1464.000000,150.000000,6.000000


Okay, so now let's looking at information that we want to extract from the dataframe. Let's say I wanted to know the max value of a certain column. The function **max()** will show you the maximum values of all columns

In [ ]:
df.max()

,0
Season,2016
Daynum,132
Wteam,1464
Wscore,186
Lteam,1464
Lscore,150
Wloc,N
Numot,6


Then, if you'd like to specifically get the max value for a particular column, you pass in the name of the column using the bracket indexing operator

In [ ]:
df['Wscore'].max()

186

If you'd like to find the mean of the Losing teams' score.

In [ ]:
float(df['Lscore'].mean())

64.49700940883343

But what if that's not enough? Let's say we want to actually see the game(row) where this max score happened. We can call the **argmax()** function to identify the row index

In [ ]:
int(df['Wscore'].argmax())

24970

One of the most useful functions that you can call on certain columns in a dataframe is the **value_counts()** function. It shows how many times each item appears in the column. This particular command shows the number of games in each season

In [ ]:
df['Season'].value_counts()

,count
Season,
2016,5369
2014,5362
2015,5354
2013,5320
2010,5263
2012,5253
2009,5249
2011,5246
2008,5163


**Q**: How many unique seasons are there in the dataset? Use the nunique() function.

In [ ]:
df['Season'].nunique()

32

**Q**: Find the team with the most wins. Use the value_counts() function on the Wteam column.

In [ ]:
df['Wteam'].value_counts().head(1)

,count
Wteam,
1181,819


# Acessing Values

Then, in order to get attributes about the game, we need to use the **iloc[]** function. Iloc is definitely one of the more important functions. The main idea is that you want to use it whenever you have the integer index of a certain row that you want to access. As per Pandas documentation, iloc is an "integer-location based indexing for selection by position."

In [ ]:
df.iloc[[df['Wscore'].argmax()]]

,Season,Daynum,Wteam,Wscore,Lteam,Lscore,Wloc,Numot
24970,1991,68,1258,186,1109,140,H,0


Let's take this a step further. Let's say you want to know the game with the highest scoring winning team (this is what we just calculated), but you then want to know how many points the losing team scored.

In [ ]:
df.iloc[[df['Wscore'].argmax()]]['Lscore']

,Lscore
24970,140


When you see data displayed in the above format, you're dealing with a Pandas **Series** object, not a dataframe object.

In [ ]:
type(df.iloc[[df['Wscore'].argmax()]]['Lscore'])

pandas.core.series.Series

In [ ]:
type(df.iloc[[df['Wscore'].argmax()]])

pandas.core.frame.DataFrame

The following is a summary of the 3 data structures in Pandas (Haven't ever really used Panels yet)

![](DataStructures.png)

When you want to access values in a Series, you'll want to just treat the Series like a Python dictionary, so you'd access the value according to its key (which is normally an integer index)

In [ ]:
int(df.iloc[[df['Wscore'].argmax()]]['Lscore'][24970])

140

The other really important function in Pandas is the **loc** function. Contrary to iloc, which is an integer based indexing, loc is a "Purely label-location based indexer for selection by label". Since all the games are ordered from 0 to 145288, iloc and loc are going to be pretty interchangable in this type of dataset

In [ ]:
df.iloc[:3]

,Season,Daynum,Wteam,Wscore,Lteam,Lscore,Wloc,Numot
0,1985,20,1228,81,1328,64,N,0
1,1985,25,1106,77,1354,70,H,0
2,1985,25,1112,63,1223,56,H,0


In [ ]:
df.loc[:3]

,Season,Daynum,Wteam,Wscore,Lteam,Lscore,Wloc,Numot
0,1985,20,1228,81,1328,64,N,0
1,1985,25,1106,77,1354,70,H,0
2,1985,25,1112,63,1223,56,H,0
3,1985,25,1165,70,1432,54,H,0


Notice the slight difference in that iloc is exclusive of the second number, while loc is inclusive.

Below is an example of how you can use loc to acheive the same task as we did previously with iloc

In [ ]:
int(df.loc[df['Wscore'].argmax(), 'Lscore'])

140

A faster version uses the **at()** function. At() is really useful wheneever you know the row label and the column label of the particular value that you want to get.

In [ ]:
int(df.at[df['Wscore'].argmax(), 'Lscore'])

140

If you'd like to see more discussion on how loc and iloc are different, check out this great Stack Overflow post: http://stackoverflow.com/questions/31593201/pandas-iloc-vs-ix-vs-loc-explanation. Just remember that **iloc looks at position** and **loc looks at labels**. Loc becomes very important when your row labels aren't integers.

# Sorting

Let's say that we want to sort the dataframe in increasing order for the scores of the losing team

In [ ]:
df.sort_values('Lscore').head()

,Season,Daynum,Wteam,Wscore,Lteam,Lscore,Wloc,Numot
100027,2008,66,1203,49,1387,20,H,0
49310,1997,66,1157,61,1204,21,H,0
89021,2006,44,1284,41,1343,21,A,0
85042,2005,66,1131,73,1216,22,H,0
103660,2009,26,1326,59,1359,22,H,0


**Q**: Make three dataframes that are sorted by season, winning team, and winning score respectively. Then, Using iloc, select the rows from index 100 to 200 and the columns for season, winning team, and winning score, respectively.

In [ ]:
df_season = df.sort_values('Season')
df_wteam = df.sort_values('Wteam')
df_wscore = df.sort_values('Wscore')

In [ ]:
subset_season = df_season.iloc[100:201][['Season', 'Wteam', 'Wscore']]
subset_wteam  = df_wteam.iloc[100:201][['Season', 'Wteam', 'Wscore']]
subset_wscore = df_wscore.iloc[100:201][['Season', 'Wteam', 'Wscore']]

In [ ]:
print(subset_season)
print(subset_wteam)
print(subset_wscore)

     Season  Wteam  Wscore
187    1985   1156     112
186    1985   1155      79
185    1985   1154      93
184    1985   1151      71
183    1985   1149      75
..      ...    ...     ...
283    1985   1376      94
282    1985   1373      71
281    1985   1368      91
280    1985   1360      65
279    1985   1356      71

[101 rows x 3 columns]
        Season  Wteam  Wscore
101650    2008   1102      72
115469    2011   1102      81
97721     2008   1102      67
12448     1988   1102      57
136264    2015   1102      66
...        ...    ...     ...
70112     2002   1102      70
91416     2006   1102      62
4564      1986   1102      56
125135    2013   1102      65
3096      1985   1102      52

[101 rows x 3 columns]
        Season  Wteam  Wscore
126840    2013   1366      43
11847     1988   1193      43
99013     2008   1460      43
9889      1987   1382      43
122700    2012   1349      43
...        ...    ...     ...
70987     2002   1139      45
128212    2013   1379      4

**Q**: From these three subsets you obtained above, find the season and winning team for the game with the highest winning score.

In [ ]:
best_game = subset_wscore.loc[subset_wscore['Wscore'].idxmax()]

print(f"Season: {best_game['Season']}")
print(f"Winning Team: {best_game['Wteam']}")
print(f"Winning Score: {best_game['Wscore']}")

Season: 2014
Winning Team: 1296
Winning Score: 45


# Filtering Rows Conditionally

Now, let's say we want to find all of the rows that satisy a particular condition. For example, I want to find all of the games where the winning team scored more than 150 points. The idea behind this command is you want to access the column 'Wscore' of the dataframe df (df['Wscore']), find which entries are above 150 (df['Wscore'] > 150), and then returns only those specific rows in a dataframe format (df[df['Wscore'] > 150]).

In [ ]:
df[df['Wscore'] > 150]

,Season,Daynum,Wteam,Wscore,Lteam,Lscore,Wloc,Numot
5269,1986,75,1258,151,1109,107,H,0
12046,1988,40,1328,152,1147,84,H,0
12355,1988,52,1328,151,1173,99,N,0
16040,1989,40,1328,152,1331,122,H,0
16853,1989,68,1258,162,1109,144,A,0
17867,1989,92,1258,181,1109,150,H,0
19653,1990,30,1328,173,1109,101,H,0
19971,1990,38,1258,152,1109,137,A,0
20022,1990,40,1116,166,1109,101,H,0
22145,1990,97,1258,157,1362,115,H,0


This also works if you have multiple conditions. Let's say we want to find out when the winning team scores more than 150 points and when the losing team scores below 100.

In [ ]:
df[(df['Wscore'] > 150) & (df['Lscore'] < 100)]

,Season,Daynum,Wteam,Wscore,Lteam,Lscore,Wloc,Numot
12046,1988,40,1328,152,1147,84,H,0
12355,1988,52,1328,151,1173,99,N,0
25656,1991,84,1106,151,1212,97,H,0
28687,1992,54,1261,159,1319,86,H,0
35023,1993,112,1380,155,1341,91,A,0
52600,1998,33,1395,153,1410,87,H,0


**Q**: Create a new column in the DataFrame called 'ScoreDifference' which is the absolute difference between the winning score and the losing score. Filter the DataFrame to only include games where the 'ScoreDifference' is greater than the average 'ScoreDifference' for all games.

In [ ]:
df['ScoreDifference'] = abs(df['Wscore'] - df['Lscore'])
avg_score_diff = df['ScoreDifference'].mean()
print(f"Average ScoreDifference: {avg_score_diff:.2f}")
df_filtered = df[df['ScoreDifference'] > avg_score_diff]
print(f"Antal rækker i original dataframe: {len(df)}")
print(f"Antal rækker efter filtrering: {len(df_filtered)}")
df_filtered.head()

Average ScoreDifference: 12.10
Antal rækker i original dataframe: 145289
Antal rækker efter filtrering: 57227


,Season,Daynum,Wteam,Wscore,Lteam,Lscore,Wloc,Numot,ScoreDifference
0,1985,20,1228,81,1328,64,N,0,17
3,1985,25,1165,70,1432,54,H,0,16
6,1985,25,1228,64,1226,44,N,0,20
8,1985,25,1260,98,1133,80,H,0,18
10,1985,25,1307,103,1288,71,H,0,32


**Q**: From this filtered DataFrame, find the season and teams involved in the game with the highest 'ScoreDifference'.

In [ ]:
best_game = df_filtered.loc[df_filtered['ScoreDifference'].idxmax()]
print(f"Season: {best_game['Season']}")
print(f"Winning Team: {best_game['Wteam']}")
print(f"Losing Team: {best_game['Lteam']}")
print(f"Score Difference: {best_game['ScoreDifference']}")

Season: 1996
Winning Team: 1409
Losing Team: 1341
Score Difference: 91


# Grouping

Another important function in Pandas is **groupby()**. This is a function that allows you to group entries by certain attributes (e.g Grouping entries by Wteam number) and then perform operations on them. The following function groups all the entries (games) with the same Wteam number and finds the mean for each group.

In [ ]:
df.groupby('Wteam')['Wscore'].mean().head()

,Wscore
Wteam,
1101,78.111111
1102,69.893204
1103,75.839768
1104,75.825944
1105,74.960894


This next command groups all the games with the same Wteam number and finds where how many times that specific team won at home, on the road, or at a neutral site

In [ ]:
df.groupby('Wteam')['Wloc'].value_counts().head(9)

Wteam  Wloc
1101   H        12
       A         3
       N         3
1102   H       204
       A        73
       N        32
1103   H       324
       A       153
       N        41
Name: count, dtype: int64

Each dataframe has a **values** attribute which is useful because it basically displays your dataframe in a numpy array style format

In [ ]:
df.values

array([[1985, 20, 1228, ..., 'N', 0, 17],
       [1985, 25, 1106, ..., 'H', 0, 7],
       [1985, 25, 1112, ..., 'H', 0, 7],
       ...,
       [2016, 132, 1246, ..., 'N', 1, 5],
       [2016, 132, 1277, ..., 'N', 0, 4],
       [2016, 132, 1386, ..., 'N', 0, 13]], dtype=object)

Now, you can simply just access elements like you would in an array.

In [ ]:
df.values[0][0]

1985

**Q**: Group the DataFrame by season and find the average winning score for each season.

In [ ]:
df.groupby('Season')['Wscore'].mean()

,Wscore
Season,
1985,74.723040
1986,74.813640
1987,77.993870
1988,79.773704
1989,81.728511
1990,80.846477
1991,82.733932
1992,79.992004
1993,79.508790


**Q**: Group the DataFrame by winning team and find the maximum winning score for each team across all seasons.

In [ ]:
df.groupby(['Season', 'Wteam'])['Wscore'].mean()

Season  Wteam
1985    1102     71.000000
        1103     70.222222
        1104     72.095238
        1106     75.100000
        1108     85.842105
                   ...    
2016    1460     73.450000
        1461     73.416667
        1462     83.222222
        1463     78.238095
        1464     85.888889
Name: Wscore, Length: 10172, dtype: float64

**Q**: Group the DataFrame by both season and winning team. Find the team with the highest average winning score for each season.

In [ ]:
best_team_per_season = df.groupby(['Season', 'Wteam'])['Wscore'].mean()

best_team_per_season = best_team_per_season.groupby('Season').idxmax()

print(best_team_per_season)

Season
1985    (1985, 1328)
1986    (1986, 1109)
1987    (1987, 1380)
1988    (1988, 1258)
1989    (1989, 1258)
1990    (1990, 1258)
1991    (1991, 1380)
1992    (1992, 1380)
1993    (1993, 1380)
1994    (1994, 1380)
1995    (1995, 1206)
1996    (1996, 1206)
1997    (1997, 1254)
1998    (1998, 1395)
1999    (1999, 1317)
2000    (2000, 1395)
2001    (2001, 1395)
2002    (2002, 1242)
2003    (2003, 1395)
2004    (2004, 1194)
2005    (2005, 1311)
2006    (2006, 1190)
2007    (2007, 1440)
2008    (2008, 1440)
2009    (2009, 1377)
2010    (2010, 1440)
2011    (2011, 1449)
2012    (2012, 1370)
2013    (2013, 1377)
2014    (2014, 1322)
2015    (2015, 1322)
2016    (2016, 1146)
Name: Wscore, dtype: object


**Q**: Create a new DataFrame that counts the number of wins for each team in each season. This will involve grouping by both season and winning team, and then using the count() function.

In [ ]:
wins_per_team = df.groupby(['Season', 'Wteam'])['Wscore'].count().reset_index()
wins_per_team.columns = ['Season', 'Wteam', 'Wins']
print(wins_per_team)

       Season  Wteam  Wins
0        1985   1102     5
1        1985   1103     9
2        1985   1104    21
3        1985   1106    10
4        1985   1108    19
...       ...    ...   ...
10167    2016   1460    20
10168    2016   1461    12
10169    2016   1462    27
10170    2016   1463    21
10171    2016   1464     9

[10172 rows x 3 columns]


**Q**: For each season, find the team with the most wins. This will involve creating a DataFrame similar to the one in task 5, and then using the idxmax() function for each season.

In [ ]:
most_wins = wins_per_team.groupby('Season')['Wins'].idxmax()
most_wins = wins_per_team.loc[most_wins][['Season', 'Wteam', 'Wins']]
print(most_wins)

      Season  Wteam  Wins
217     1985   1385    27
342     1986   1181    32
818     1987   1424    33
861     1988   1112    31
1323    1989   1328    28
1551    1990   1247    29
1739    1991   1116    30
2088    1992   1181    28
2423    1993   1231    28
2665    1994   1163    27
2928    1995   1116    26
3357    1996   1269    31
3638    1997   1242    31
3943    1998   1242    33
4201    1999   1181    32
4709    2000   1409    29
4829    2001   1181    29
5121    2002   1153    30
5456    2003   1166    29
5981    2004   1390    29
6165    2005   1228    32
6452    2006   1181    30
6841    2007   1242    30
7209    2008   1272    33
7552    2009   1272    31
7870    2010   1242    32
8217    2011   1242    32
8565    2012   1246    32
8878    2013   1211    30
9462    2014   1455    33
9609    2015   1246    34
9955    2016   1242    29


**Q**: Group the DataFrame by losing team and find the average losing score for each team across all seasons. Compare this with the average winning score for each team from task 3. Are there teams that have a higher average losing score than winning score?

In [ ]:
avg_losing = df.groupby('Lteam')['Lscore'].mean()
avg_winning = df.groupby('Wteam')['Wscore'].mean()

comparison = pd.DataFrame({
    'avg_losing_score': avg_losing,
    'avg_winning_score': avg_winning
}).dropna()

higher_losing = comparison[comparison['avg_losing_score'] > comparison['avg_winning_score']]

if len(higher_losing) == 0:
    print("Ingen hold har højere gennemsnitlig taberscore end vinderscore")
else:
    print(higher_losing)

Ingen hold har højere gennemsnitlig taberscore end vinderscore


# Dataframe Iteration

In order to iterate through dataframes, we can use the **iterrows()** function. Below is an example of what the first two rows look like. Each row in iterrows is a Series object

In [ ]:
for index, row in df.iterrows():
    print(row)
    if index == 1:
        break

Season             1985
Daynum               20
Wteam              1228
Wscore               81
Lteam              1328
Lscore               64
Wloc                  N
Numot                 0
ScoreDifference      17
Name: 0, dtype: object
Season             1985
Daynum               25
Wteam              1106
Wscore               77
Lteam              1354
Lscore               70
Wloc                  H
Numot                 0
ScoreDifference       7
Name: 1, dtype: object


**Q**: Create a new column 'HighScoringGame' that is 'Yes' if the winning score is greater than 100 and 'No' otherwise. This will require iterating over the rows of the DataFrame and checking the value of the winning score for each row.

In [ ]:
for index, row in df.iterrows():
    if row['Wscore'] > 100:
        df.at[index, 'HighScoringGame'] = 'Yes'
    else:
        df.at[index, 'HighScoringGame'] = 'No'

print(df[['Wscore', 'HighScoringGame']].head(10))

   Wscore HighScoringGame
0      81              No
1      77              No
2      63              No
3      70              No
4      86              No
5      79              No
6      64              No
7      58              No
8      98              No
9      97              No


**Q**: Calculate the total number of games played by each team, whether they won or lost. This will require iterating over the rows of the DataFrame and updating a dictionary that keeps track of the number of games for each team.

In [ ]:
games_played = {}

for index, row in df.iterrows():
    wteam = row['Wteam']
    lteam = row['Lteam']

    if wteam not in games_played:
        games_played[wteam] = 0
    if lteam not in games_played:
        games_played[lteam] = 0

    games_played[wteam] += 1
    games_played[lteam] += 1

print(pd.Series(games_played).sort_values(ascending=False).head(10))

1181    1013
1314    1010
1246     999
1257     994
1242     993
1393     993
1272     992
1228     992
1437     983
1400     982
dtype: int64


**Q**: For each season, find the game with the highest score difference (winning score - losing score). This will require iterating over the rows of the DataFrame, keeping track of the highest score difference for each season, and updating it if a game with a higher score difference is found.

In [ ]:
highest_diff = {}

for index, row in df.iterrows():
    season = row['Season']
    diff = row['Wscore'] - row['Lscore']

    if season not in highest_diff:
        highest_diff[season] = {'diff': diff, 'index': index}

    if diff > highest_diff[season]['diff']:
        highest_diff[season] = {'diff': diff, 'index': index}

for season, game in sorted(highest_diff.items()):
    row = df.loc[game['index']]
    print(f"Season {int(season)}: Wteam {int(row['Wteam'])} vs Lteam {int(row['Lteam'])}, Score Difference: {int(game['diff'])}")

Season 1985: Wteam 1361 vs Lteam 1288, Score Difference: 60
Season 1986: Wteam 1314 vs Lteam 1264, Score Difference: 84
Season 1987: Wteam 1155 vs Lteam 1118, Score Difference: 73
Season 1988: Wteam 1328 vs Lteam 1147, Score Difference: 68
Season 1989: Wteam 1242 vs Lteam 1135, Score Difference: 70
Season 1990: Wteam 1181 vs Lteam 1217, Score Difference: 76
Season 1991: Wteam 1163 vs Lteam 1148, Score Difference: 68
Season 1992: Wteam 1116 vs Lteam 1126, Score Difference: 82
Season 1993: Wteam 1328 vs Lteam 1197, Score Difference: 81
Season 1994: Wteam 1228 vs Lteam 1152, Score Difference: 69
Season 1995: Wteam 1246 vs Lteam 1404, Score Difference: 74
Season 1996: Wteam 1409 vs Lteam 1341, Score Difference: 91
Season 1997: Wteam 1278 vs Lteam 1106, Score Difference: 80
Season 1998: Wteam 1395 vs Lteam 1410, Score Difference: 66
Season 1999: Wteam 1268 vs Lteam 1317, Score Difference: 75
Season 2000: Wteam 1261 vs Lteam 1212, Score Difference: 75
Season 2001: Wteam 1196 vs Lteam 1197, S

Remember, iterating over a DataFrame should generally be avoided if a vectorized operation can be used instead, as vectorized operations are usually much faster. However, these tasks are designed to give practice with DataFrame iteration for cases where it might be necessary.

Vectorized Operation Example: Create a new column 'HighScoringGame' in the DataFrame using a vectorized operation. This column should contain 'Yes' if the winning score is greater than 100 and 'No' otherwise. Use the np.where function from the numpy library for this task.

In [ ]:
import numpy as np
df['HighScoringGame'] = np.where(df['Wscore'] > 100, 'Yes', 'No')

**Q**: Vectorized Operation: Calculate the total number of games played by each team, whether they won or lost. Instead of iterating over the DataFrame, use the value_counts() function on the winning team and losing team columns separately, and then add the two Series together.

In [ ]:
wins = df['Wteam'].value_counts()
losses = df['Lteam'].value_counts()

total_games = wins.add(losses, fill_value=0).astype(int)

print(total_games.sort_values(ascending=False).head(10))

1181    1013
1314    1010
1246     999
1257     994
1393     993
1242     993
1272     992
1228     992
1437     983
1400     982
Name: count, dtype: int64


**Q**: For each season, find the game with the highest score difference (winning score - losing score). Instead of iterating over the DataFrame, create a new column 'ScoreDifference' using vectorized subtraction, then use the groupby() function and idxmax() function to find the game with the highest score difference for each season.

In [ ]:
df['ScoreDifference'] = df['Wscore'] - df['Lscore']

idx = df.groupby('Season')['ScoreDifference'].idxmax()

result = df.loc[idx][['Season', 'Wteam', 'Lteam', 'ScoreDifference']]
print(result)

        Season  Wteam  Lteam  ScoreDifference
236       1985   1361   1288               60
4731      1986   1314   1264               84
8240      1987   1155   1118               73
12046     1988   1328   1147               68
16677     1989   1242   1135               70
19502     1990   1181   1217               76
25161     1991   1163   1148               68
27997     1992   1116   1126               82
33858     1993   1328   1197               81
36404     1994   1228   1152               69
39858     1995   1246   1404               74
44653     1996   1409   1341               91
49033     1997   1278   1106               80
52600     1998   1395   1410               66
57618     1999   1268   1317               75
60616     2000   1261   1212               75
65999     2001   1196   1197               75
70737     2002   1328   1183               59
74791     2003   1403   1311               72
79181     2004   1328   1115               70
83372     2005   1403   1421      

# Extracting Rows and Columns

The bracket indexing operator is one way to extract certain columns from a dataframe.

In [ ]:
df[['Wscore', 'Lscore']].head()

,Wscore,Lscore
0,81,64
1,77,70
2,63,56
3,70,54
4,86,74


Notice that you can acheive the same result by using the loc function. Loc is a veryyyy versatile function that can help you in a lot of accessing and extracting tasks.

In [ ]:
df.loc[:, ['Wscore', 'Lscore']].head()

,Wscore,Lscore
0,81,64
1,77,70
2,63,56
3,70,54
4,86,74


Note the difference is the return types when you use brackets and when you use double brackets.

In [ ]:
type(df['Wscore'])

pandas.core.series.Series

In [ ]:
type(df[['Wscore']])

pandas.core.frame.DataFrame

You've seen before that you can access columns through df['col name']. You can access rows by using slicing operations.

In [ ]:
df[0:3]

,Season,Daynum,Wteam,Wscore,Lteam,Lscore,Wloc,Numot,ScoreDifference,HighScoringGame
0,1985,20,1228,81,1328,64,N,0,17,No
1,1985,25,1106,77,1354,70,H,0,7,No
2,1985,25,1112,63,1223,56,H,0,7,No


Here's an equivalent using iloc

In [ ]:
df.iloc[0:3,:]

,Season,Daynum,Wteam,Wscore,Lteam,Lscore,Wloc,Numot,ScoreDifference,HighScoringGame
0,1985,20,1228,81,1328,64,N,0,17,No
1,1985,25,1106,77,1354,70,H,0,7,No
2,1985,25,1112,63,1223,56,H,0,7,No


# Data Cleaning

One of the big jobs of doing well in Kaggle competitions is that of data cleaning. A lot of times, the CSV file you're given (especially like in the Titanic dataset), you'll have a lot of missing values in the dataset, which you have to identify. The following **isnull** function will figure out if there are any missing values in the dataframe, and will then sum up the total for each column. In this case, we have a pretty clean dataset.

In [ ]:
df.isnull().sum()

,0
Season,0
Daynum,0
Wteam,0
Wscore,0
Lteam,0
Lscore,0
Wloc,0
Numot,0
ScoreDifference,0
HighScoringGame,0


If you do end up having missing values in your datasets, be sure to get familiar with these two functions.
* **dropna()** - This function allows you to drop all(or some) of the rows that have missing values.
* **fillna()** - This function allows you replace the rows that have missing values with the value that you pass in.

# Other Useful Functions

* **drop()** - This function removes the column or row that you pass in (You also have the specify the axis).
* **agg()** - The aggregate function lets you compute summary statistics about each group
* **apply()** - Lets you apply a specific function to any/all elements in a Dataframe or Series
* **get_dummies()** - Helpful for turning categorical data into one hot vectors.
* **drop_duplicates()** - Lets you remove identical rows

# Lots of Other Great Resources

Pandas has been around for a while and there are a lot of other good resources if you're still interested on getting the most out of this library.
* http://pandas.pydata.org/pandas-docs/stable/10min.html
* https://www.datacamp.com/community/tutorials/pandas-tutorial-dataframe-python
* http://www.gregreda.com/2013/10/26/intro-to-pandas-data-structures/
* https://www.dataquest.io/blog/pandas-python-tutorial/
* https://drive.google.com/file/d/0ByIrJAE4KMTtTUtiVExiUGVkRkE/view
* https://www.youtube.com/playlist?list=PL5-da3qGB5ICCsgW1MxlZ0Hq8LL5U3u9y